rados/read_op_omap_get_vals_by_keys.go

```diff
@@ -0,0 +1,128 @@
+//go:build ceph_preview
+// +build ceph_preview
+
+package rados
+
+// #cgo LDFLAGS: -lrados
+// #include <rados/librados.h>
+// #include <stdlib.h>
+//
+import "C"
+
+import (
+	"unsafe"
+
+	"github.com/ceph/go-ceph/internal/cutil"
+)
+
+// ReadOpOmapGetValsByKeysStep holds the result of the
+// GetOmapValuesByKeys read operation.
+// Result is valid only after Operate() was called.
+type ReadOpOmapGetValsByKeysStep struct {
+	withRefs
+
+	// C arguments
+
+	cKeys cutil.CPtrCSlice
+	iter  C.rados_omap_iter_t
+	prval *C.int
+
+	// Internal state
+
+	// canIterate is only set after the operation is performed and is
+	// intended to prevent premature fetching of data
+	canIterate bool
+}
+
+func newReadOpOmapGetValsByKeysStep(keys []string) *ReadOpOmapGetValsByKeysStep {
+	s := &ReadOpOmapGetValsByKeysStep{
+		cKeys: cutil.NewCPtrCSlice(len(keys)),
+		prval: (*C.int)(C.malloc(C.sizeof_int)),
+	}
+
+	for i, key := range keys {
+		cKey := C.CString(key)
+		s.add(unsafe.Pointer(cKey))
+		s.cKeys[i] = cutil.CPtr(cKey)
+	}
+
+	return s
+}
+
+func (s *ReadOpOmapGetValsByKeysStep) free() {
+	s.canIterate = false
+
+	if s.iter != nil {
+		C.rados_omap_get_end(s.iter)
+	}
+	s.iter = nil
+
+	s.cKeys.Free()
+	s.cKeys = nil
+	s.withRefs.free()
+	C.free(unsafe.Pointer(s.prval))
+	s.prval = nil
+}
+
+func (s *ReadOpOmapGetValsByKeysStep) update() error {
+	err := getError(*s.prval)
+	s.canIterate = (err == nil)
+
+	return err
+}
+
+// Next gets the next omap key/value pair referenced by
+// ReadOpOmapGetValsByKeysStep's internal iterator.
+// If there are no more elements to retrieve, (nil, nil) is returned.
+// May be called only after Operate() finished.
+//  PREVIEW
+func (s *ReadOpOmapGetValsByKeysStep) Next() (*OmapKeyValue, error) {
+	if !s.canIterate {
+		return nil, ErrOperationIncomplete
+	}
+
+	var (
+		cKey    *C.char
+		cVal    *C.char
+		cValLen C.size_t
+	)
+
+	ret := C.rados_omap_get_next(s.iter, &cKey, &cVal, &cValLen)
+	if ret != 0 {
+		return nil, getError(ret)
+	}
+
+	if cKey == nil {
+		// Iterator has reached the end of the list.
+		return nil, nil
+	}
+
+	return &OmapKeyValue{
+		Key:   C.GoString(cKey),
+		Value: C.GoBytes(unsafe.Pointer(cVal), C.int(cValLen)),
+	}, nil
+}
```

Comment:
There are a couple more place in `rados`'s codebase that deals with OMap iterators and this adds an unnecessary amount of boilerplate. Maybe we could factor it out into its own struct and reuse that (probably as a separate PR)?

Reply:
Sure, refactoring is generally welcome especially if it reduces the maintenance burden. Usually the only cases where I'd object to it would be cases where copy-n-paste reduces the maintenace burden or if its not really clear the code paths are the same,etc. 

For this case I see two main options:
* private helper _functions_ that perform much of the boilerplate work
* a private embedded struct type that can provide the common methods

We can do the work in a follow up PR or here in this PR, but please let us know what you are thinking so I know how deeply to review the current state in the short term.  :-)

Reply:
@phlogistonjohn thanks for the suggestions. I think doing it in a follow up PR would be cleaner, so this one should be ok to review as is.